File: Chapter 02/InsertionSort.c
#include <stdio.h>
#include <stdlib.h>

void nonDecreasingInsertionSort(int *array, int arrayLength)
{
    int i, j, key;

    for(j=1; j<arrayLength; j++) {
        key = array[j];
        i = j-1;

        while(i>=0 && array[i]>key) {
            array[i+1] = array[i];
            i = i-1;
        }

        array[i+1] = key;
    }
}

void nonIncreasingInsertionSort(int *array, int arrayLength)
{
    int i, j, key;

    for(j=1; j<arrayLength; j++) {
        key = array[j];
        i = j-1;

        while(i>=0 && array[i]<key) {
            array[i+1] = array[i];
            i = i-1;
        }

        array[i+1] = key;
    }
}

int main()
{
    int *array, arrayLength, i;

    printf("How many numbers will you enter? : ");
    scanf("%d", &arrayLength);

    array = (int*)malloc(sizeof(int) * arrayLength);

    for(i=0; i<arrayLength; i++) {
        printf("Enter %d. number : ",i+1);
        scanf("%d", &array[i]);
    }

    nonDecreasingInsertionSort(array,arrayLength);
    printf("\nYour array after the non-decreasing sorting : \n");

    for(i=0; i<arrayLength; i++) {
        printf("%d ", array[i]);
    }

    nonIncreasingInsertionSort(array,arrayLength);
    printf("\nYour array after the non-increasing sorting : \n");

    for(i=0; i<arrayLength; i++) {
        printf("%d ", array[i]);
    }

    return 0;
}

File: Chapter 02/InsertionSort.java
package ınsertionsort;

public class InsertionSort {

    public static void main(String[] args) {
        int[] A = {5,2,4,6,1,3};
        
        nonDecreasingInsertionSort(A);
        for(int i=0; i<A.length; i++)
            System.out.print(A[i] + " ");
        
        System.out.println("");
        
        nonIncreasingInsertionSort(A);
        for(int i=0; i<A.length; i++)
            System.out.print(A[i] + " ");
        
    }
    
    public static void nonDecreasingInsertionSort(int[] A) {
        int i, j, key;
        for(j=1; j<A.length; j++) {
            key = A[j];
            i=j-1;
            
            while(i>=0 && A[i]>key) {
                A[i+1] = A[i];
                i--;
            }
            A[i+1] = key;
        }
    }
    
    public static void nonIncreasingInsertionSort(int[] A) {
        int i, j, key;
        for(j=1; j<A.length; j++) {
            key = A[j];
            i=j-1;
            
            while(i>=0 && A[i]<key) {
                A[i+1] = A[i];
                i--;
            }
            A[i+1] = key;
        }
    }
}
